consumer/consumererror/error.go

```diff
@@ -0,0 +1,178 @@
+
+import (
+	"errors"
+
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// Error acts as a container for errors coming from pipeline components.
+// It may hold multiple errors from downstream components, and is designed
+// to act as a way to accumulate errors as it travels upstream in a pipeline.
+// `Error` should be obtained using `errors.As` and as a result, ideally
+// a single instance should exist in an error stack. If this is not possible,
+// the most `Error` object should be highest on the stack.
+//
+// Experimental: This API is at the early stage of development and may change without backward compatibility
+type Error struct {
+	errors []ErrorData
+}
+
+// ErrorData is intended to be used to encapsulate various information
+// that can add context to an error that occurred within a pipeline component.
+// ErrorData objects are constructed through calling `New` with the relevant
+// options to capture data around the error that occurred. It can then be pulled
+// out by an upstream component by calling `Error.Data`.
+//
+// Experimental: This API is at the early stage of development and may change without backward compatibility
+type ErrorData interface {
+	Error() string
+
+	Unwrap() error
+
+	// Returns a status code and a boolean indicating whether a status was set
+	// by the error creator.
+	HTTPStatus() (int, bool)
+
+	// Returns a status code and a boolean indicating whether a status was set
+	// by the error creator.
+	GRPCStatus() (*status.Status, bool)
+
+	// Disallow implementations outside this package to allow later extending
+	// the interface.
+	unexported()
+}
+
+type errorData struct {
+	error
+	httpStatus *int
+	grpcStatus *status.Status
+}
+
+// ErrorOption allows annotating an Error with metadata.
+type ErrorOption func(error *errorData)
+
+// New wraps an error that happened while consuming telemetry
+// and adds metadata onto it to be passed back up the pipeline.
+//
+// Experimental: This API is at the early stage of development and may change without backward compatibility
+func New(origErr error, options ...ErrorOption) error {
+	errData := &errorData{error: origErr}
+	err := &Error{errors: []ErrorData{errData}}
+
+	for _, option := range options {
+		option(errData)
+	}
+
+	return err
+}
+
+var _ error = &Error{}
+
+// Error implements the `error` interface.
+func (e *Error) Error() string {
+	return e.errors[len(e.errors)-1].Error()
+}
+
+// Unwrap returns the wrapped error for use by `errors.Is` and `errors.As`.
+func (e *Error) Unwrap() []error {
+	errors := make([]error, 0, len(e.errors))
+
+	for _, err := range e.errors {
+		errors = append(errors, err)
+	}
+
+	return errors
+}
+
+// Data returns all the accumulated ErrorData errors
+// emitted by components downstream in the pipeline.
+// These can then be aggregated or worked with individually.
+func (e *Error) Data() []ErrorData {
```

Comment:
This is wrong I think, we should not return the internal slice since the caller can modify it.

Reply:
From what I can tell, this creates a new slice instance where modifications don't affect the original slice.

I added a test here that seems to show this isn't an issue, is there anything else I should check for to ensure we don't pass out a mutable copy of the slice? https://github.com/open-telemetry/opentelemetry-collector/pull/9041/files#diff-8c40d97e3ff1a76b10c0c2b02a2cea669f1ffdffc2363cfdf86ba8489cb9f59cR92